src/pymedphys/dcm/_level1/dcmdose.py

```diff
@@ -41,12 +40,33 @@
 from ...libutils import get_imports
 IMPORTS = get_imports(globals())
 
-Coords = namedtuple('Coords', 'x y z')
-Dose = namedtuple(
-    'Dose', 'values units type summation heterogeneity_correction')
-
 # pylint: disable=C0103
 
+class DicomDose:
+
+    def __init__(self, dcm_filepath):
+        dcm = pydicom.dcmread(dcm_filepath)
+        if dcm.Modality != "RTDOSE":
+            raise ValueError("The input DICOM file is not an RT Dose file")
+        self.dcm = dcm
+
+        self.values = dcm.pixel_array * dcm.DoseGridScaling
+        self.units = dcm.DoseUnits
+        self.x, self.y, self.z = extract_dicom_patient_xyz(dcm)
+        self.coords = convert_xyz_to_dicom_coords((self.x, self.y, self.z))
```

Comment:
May add a `coord_system` attribute later, with some native function to convert `self.coords` and track the current system. Being always able to refer to `self.ds` for patient orientation.

Reply:
When this happens, my preference is that the coords transformation function returns a new DicomDose object instead of making this change you refer to inplace.

Reply:
I don't think I mind too much, but why is that? That would be less memory efficient and the original version of the object would be easily retrievable through a simple conversion back.

Basically, it seems like a mutable thing to me...

Reply:
People using the object not understanding the difference between mutable and immutable can fall into some really awkward bugs when working with something mutable and assuming it is immutable.

For example:

```python
dose_dicom1 = DoseDicom(...)
dose_dicom2 = DoseDicom(...)

collection = [dose_dicom1, dose_dicom2]

collection[0].convert_coords(something)

dose_dicom1.dosomething # (but now it has changed)
```

Reply:
I made that mistake with the standard pydicom object, that's why I needed to put a deepcopy on that template object.

With the RAM point, I imagine conversion of coordinates should only ever need to be done once or twice to an object. So shouldn't be too much of an issue here.

Reply:
I don't think that's a reason to insist on immutable objects. That exact same point could be made for many python objects that we regularly use throughout the PyMedPhys code, like lists and numpy arrays.

To be honest, in the code snippet you provided, I would *expect* `dose_dicom1` to have changed. Indeed, with code like that, I would've *intended* for it to change. I actually think the immutable case is the more confusing .

The mutable version better matches intuition. When my car has an empty fuel tank and I go to the petrol station to fill it up, I don't discard my empty-tanked car for a replica with a full tank!

> With the RAM point, I imagine conversion of coordinates should only ever need to be done once or twice to an object. So shouldn't be too much of an issue here.

Perhaps you're right, but I think it's a bit brave to assume this. I can absolutely see a batch processing application here...

Reply:
I don't think that's a reason to insist on immutable objects. That exact same point could be made for many python objects that we regularly use throughout the PyMedPhys code, like lists and numpy arrays.

The mutable version better matches intuition. When my car has an empty fuel tank and I go to the petrol station to fill it up, I don't discard my empty-tanked car for a replica with a full tank!

> With the RAM point, I imagine conversion of coordinates should only ever need to be done once or twice to an object. So shouldn't be too much of an issue here.

Perhaps you're right, but I think it's a bit brave to assume this. I can absolutely see a batch processing application here...

---

I think I might have a way to circumvent both of our concerns.

It does make sense to perform a deepcopy on the `dcm` at DicomDose construction, but then I'd still prefer to keep it mutable from there. In that way, the following isn't a risk:

```python
dcm = pydicom.dcm_read("my_dicom_file.dcm")

dcmdose = DicomDose(dcm)
dcmdose.convert_coords(...)

dcm.do_something() # but dcm has changed as a result of methods called on dcmdose
```

If `dcmdose.dcm` (i.e. the pydicom Dataset used to construct the DicomDose object) is copied before saving it as an attribute, the above will not occur (`dcm` will not have changed).

I should then be okay for methods like `dcmdose.convert_coords()` to simply edit `dcmdose.dcm` rather than return a deepcopy of `dcmdose.dcm` with the edits. There is no risk of the above confusion *and* you avoid memory waste and possible associated performance issues (e.g. in batch processing).